doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,75 +1,368 @@
 [CPROVER Manual TOC](../../)
 
-# Assigns Clause
+# Assigns Clauses
 
-## In Function Contracts
+An _assigns clause_ lets the user to specify which memory locations may be
+assigned to by a function or a loop.
 
-### Syntax
+For a function contract, if no assigns clause is provided, the default is the
+empty set.
+
+For a loop contract, if no assigns clause is provided, CBMC attempts to infer
+the set of locations assigned by the loop from the loop body,
+and then checks that the inferred set is correct (as if specified by the user).
+
+For both functions and loop contracts, if more than one assigns clause is
+provided, they are merged into a single clause.
+
+We use the _assigns_ interpretation for these memory locations, which means that
+memory locations that are part of the _frame_ of the function call
+(or the loop) and that are not listed in the assigns clause
+(or the _inferred_ assigns clause for a loop contract) must not be assigned to
+by the function (or the loop), even if they end up holding the same value as they
+held before the function call (or before entering the loop).
+
+Memory locations that are not part of the _frame_, i.e. that are locally stack-
+or heap-allocated during function execution or loop execution can always be
+assigned to by the function or the loop.
```

Comment:
I think we need to disentangle the introduction of "frame" and the difference between "modifies" and "assigns."

Reply:
removed the references to frames